js/modules/debug.js

```diff
@@ -119,17 +126,40 @@ const _createMessage = ({ commonProperties, conversationId, type } = {}) => {
   }
 };
 
-const MEGA_BYTE = 1e6;
-const createRandomInMemoryAttachment = () => {
-  const numBytes = (1 + Math.ceil((Math.random() * 50))) * MEGA_BYTE;
-  const array = new Uint32Array(numBytes).fill(1);
-  const data = array.buffer;
-  const fileName = Math.random().toString().slice(2);
+const FIXTURES_PATH = path.join(__dirname, '..', '..', 'fixtures');
+const readData = Attachments.readData(FIXTURES_PATH);
```

Comment:
This is a weird line. Maybe the right name for Attachments.readData() is something like Attachments.createLoader() or something?

Reply:
I hear you! The original intention was to use currying (technique where every function takes one argument) to allow better composition for places where we do `Array::map`, etc. For example the function `Message.upgradeSchemaVersion(message, context)` can’t easily be passed into `Array::map` because it takes two arguments (`(message, context)`) instead of just `(message)`. Using currying, e.g. `Message.upgradeSchema = context => message => {…}`, one can bind `context` once, e.g. `upgradeMessageSchemaVersion = Message.upgradeSchema(context)`, and then easily pass it into `Array::map`, e.g. `await Promise.all(messages.map(upgradeMessageSchema))`. Similar reasons apply to `readData`, e.g. 

- Lodash offers FP variant that does this (iteratee-first, data-last): https://github.com/lodash/lodash/wiki/FP-Guide#lodashfp
- Ramda.js is built on this principle too: http://ramdajs.com/

I agree, here in vanilla JS and without type checking it’s not clear that `readData` returns a function that takes a path vs directly reading the data. I’ll update it to `createReader`, `createWriter`, and `createDeleter`, but it will cause some churn and risk because it’ll be hard to test all the different code paths. As we adopt TypeScript, I hope we can leverage this powerful technique with more confidence as the compiler will assist us in making sure we correctly compose functions 😄 

Let me know if you have any questions or concerns.

Reply:
Yep, I'e used lodash's 'fp' function variants for some very cool composition. But as you mention, currying can be extremely confusing without types, unless you use very clear names, or make it a very common pattern in the code.

Of course, this may be a special case. `const something = something()` is probably confusing no matter how you slice it.

Reply:
Cool, I’ve always been wanting to use more of `lodash/fp` but as you said, it’s hard without types! 

This article talks about that more in-depth:
https://www.reaktor.com/blog/fear-trust-and-javascript/

I look forward to exploring this area more in the future 😄